### ISSUE TITLE
Improve code folding for decorators

### ISSUE THREAD

**Issue by Geo5** [astral-sh/ty#2861]
### Summary

Hi, thank you for the work!
I noticed some potential improvements for the (very) recently implemented code folding (#2588) regarding function decorators:
ty currently only creates one fold including the decorator(s) and the function, which leads to this:
```python
class Foo:
    @staticmethod
    def bar(arg: int) -> None:
        pass
```
<img width="528" height="170" alt="Image" src="https://github.com/user-attachments/assets/56f9fe7e-6cac-4c08-a782-b149f593dd0b" />

Which i think is not the best possible experience, because it hides the actual function name.

If i disable language services, the native VSCode folding looks like this, with 2 levels of folding for decorators and the function itself:

<img width="611" height="158" alt="Image" src="https://github.com/user-attachments/assets/62de1cd9-244c-4a9c-a00d-e4b3c1777316" />

<img width="762" height="195" alt="Image" src="https://github.com/user-attachments/assets/37700ba6-e54b-4234-a97f-6bafcae4654b" />

Which i like more, because i can see function names without unfolding the implementation. I think there could also be an argument for not creating a fold for the decorators at all, or at least not if there is only one decorator, but i am not sure what i would like best there.

**Comment by iksuddle** [comment]
Hey, I’ve been looking into it and would appreciate some feedback on my approach:

After parsing decorators in the parser's `parse_decorators`, we can get the start of the definition using `self.node_start()`.

From there, one option would be to pass that as the `start` argument for `parse_function_definition` / `parse_class_definition`, or alternatively pass it as an additional argument while creating a new field in `ast::StmtFunctionDef` / `ast::StmtClassDef`. If decorators don’t need to be included in the range, I think the first approach is better. Please let me know if I'm missing something.

With this, `FoldingRangeVisitor::enter_node` should fold from the actual function/class definition rather than the first decorator.

If folding decorators is also needed, the range for that could also be recorded in `FoldingRangeVisitor::enter_node` by accessing the function/class `decorator_list`.

**Comment by denyszhak** [comment]
@iksuddle 
If I understood your proposed approach right, changing `StmtFunctionDef` or `StmtClassDef` ranges in the parser feels broader and could affect AST behaviour outside folding. Since this issue is in ty server folding, it seems safer to keep it in `FoldingRangeVisitor`.

I may be wrong here since I'm not a maintainer, just sharing this in case it helps before you submit a change.

**Comment by dhruvmanila** [comment]
Hmm, it does seem that the CPython parser does _not_ include the decorator as part of the function range. For example, in the following code:

```py
class Foo:                                                                                                    
    @staticmethod                                                                                             
    def bar(arg: int) -> None:                                                                                
        pass   
```

I get:

```
FunctionDef: lineno=3, col_offset=4, end_lineno=4, end_col_offset=12
decorator: lineno=2, col_offset=5, end_lineno=2, end_col_offset=17
```

Notice how the decorator starts from line 2 while the function starts from line 3.

But, I agree with @denyszhak that we should keep the fix for this specific issue local for now as changing the AST range will probably have downstream effects on the linter, formatter, etc.

**Comment by dhruvmanila** [comment]
Claude found the PR where this behavior was changed https://github.com/astral-sh/ruff/pull/4467 and the referenced parser PR https://github.com/astral-sh/RustPython-Parser/pull/2, so this is intentional.

We should fix this in the `ty_server` instead.

**Comment by MichaReiser** [comment]
This is how we handle this in the formatter

https://github.com/astral-sh/ruff/blob/62343a101a44bcd5acfd23a55beb8b2e7f701809/crates/ruff_python_formatter/src/statement/clause.rs#L249-L271

I think it makes sense to add helper methods to both the `function` and `class` AST nodes that return the range without decorators

**Comment by dhruvmanila** [comment]
> I think it makes sense to add helper methods to both the `function` and `class` AST nodes that return the range without decorators

For now, [the PR](https://github.com/astral-sh/ruff/pull/23543) makes use of the `Tokens` struct which contains helper function to find the token that starts the function / class definition. I see that the formatter uses the `SimpleTokenizer` to find it. I guess adding these methods to the AST node does sound better.
